client/core/exchangeratefetcher.go

```diff
@@ -0,0 +1,276 @@
+	}
+	return nil
+}
+
+// Used to initialize the embedding sources.
+func newCommonFetcher(token string, logger dex.Logger) *commonFetcher {
+	return &commonFetcher{
+		token:      token,
+		client:     new(http.Client),
+		log:        logger,
+		assetRates: make(map[uint32]*AssetRateInfo),
+	}
+}
+
+type coinPaprikaSource struct {
+	*commonFetcher
+}
+
+// coinPaprikaResponse models the JSON data returned from the coinPaprika API.
+type coinPaprikaResponse struct {
+	Name   string `json:"name"`
+	Symbol string `json:"symbol"`
+	Quotes struct {
+		Currency struct {
+			Price float64 `json:"price"`
+		} `json:"USD"`
+	} `json:"quotes"`
+}
+
+// newCoinPaprika constructs a coinPaprikaSource.
+func newCoinPaprika(logger dex.Logger) ExchangeRateFetcher {
+	return &coinPaprikaSource{
+		commonFetcher: newCommonFetcher(coinPaprika, logger),
+	}
+}
+
+// FetchRates retrieves and parses exchange rate data from API.
+func (fetcher *coinPaprikaSource) FetchRates(assests map[uint32]*SupportedAsset) {
+	fetcher.logRequest()
+	for assetID, asset := range assests {
+		if asset.Wallet == nil {
+			// we don't want to fetch rates for assets with no wallet.
+			continue
+		}
+		res := new(coinPaprikaResponse)
+		slug := fmt.Sprintf("%s-%s", asset.Symbol, asset.Info.Name)
+		// Special handling for asset names with multiple space, e.g Bitcoin Cash.
+		slug = strings.ToLower(strings.ReplaceAll(slug, " ", "-"))
+		reqStr := fmt.Sprintf(coinPaprikaURL, slug)
+		request, err := http.NewRequest(http.MethodGet, reqStr, nil)
+		if err != nil {
+			fetcher.log.Errorf("%s: NewRequest %v\n", fetcher.token, err)
+			continue
+		}
+		err = fetcher.fetch(request, res)
+		if err != nil {
+			fetcher.log.Errorf("%s: Fetch %v\n", fetcher.token, err)
+			continue
+		}
+		fetcher.updateRate(assetID, res.Quotes.Currency.Price)
+	}
+}
+
+type dcrDataSource struct {
+	*commonFetcher
+}
+
+type dcrdataResponse struct {
+	Currency string  `json:"btcIndex"`
+	DcrPrice float64 `json:"dcrPrice"`
+	BtcPrice float64 `json:"btcPrice"`
+}
+
+// newDcrdata constructs a dcrDataSource.
+func newDcrdata(logger dex.Logger) ExchangeRateFetcher {
+	return &dcrDataSource{
+		commonFetcher: newCommonFetcher(dcrataDotOrg, logger),
+	}
+}
+
+// FetchRates retrieves and parses exchange rate data from API.
+func (fetcher *dcrDataSource) FetchRates(assets map[uint32]*SupportedAsset) {
+	fetcher.logRequest()
+	assetBTC := assets[btcBipId]
+	assetDCR := assets[dcrBipId]
+	if (assetBTC == nil || assetBTC.Wallet == nil) && (assetDCR == nil || assetDCR.Wallet == nil) {
+		return
+	}
+	res := new(dcrdataResponse)
+	request, err := http.NewRequest(http.MethodGet, dcrDataURL, nil)
+	if err != nil {
+		fetcher.log.Errorf("%s: NewRequest %v\n", fetcher.token, err)
+		return
+	}
+	err = fetcher.fetch(request, res)
+	if err != nil {
+		fetcher.log.Errorf("%s: Fetch %v\n", fetcher.token, err)
+		return
+	}
+
+	fetcher.updateRate(btcBipId, res.BtcPrice)
+	fetcher.updateRate(dcrBipId, res.DcrPrice)
```

Comment:
I guess it's possible the client only has one of these wallets. Could not do both in that case.